### PR TITLE
WIP: Merge Sub-Projects

### LINKED ISSUE #44: Merge Sub-Projects
the separation into `core` and `desktop`, even if it appears elegant and future-proof, does add a certain complexity that does not seem justified to me at the moment (and for the foreseeable future). next year, when we have students work directly with this repo as a default, this could^Wwill become another possible source of potential problems. 

should we merge `core` and `desktop` for the time being?

this would also affect build skripts (gradle, maven) and maybe documentation.

### PR DESCRIPTION
**Pull request by Tobias321**
resolves #44 

works fine for me :)

edit: i cloned this branch again. now you can do two things:

- import an existing gradle project in eclipse,
- run the gradle task `jar` to build a jar with all dependencies in `build/libs/pmdungeon-1.0.jar` **or**
- add a Main class and copy the assets folder, and then run the project in eclipse:

```java
package de.pmdungeon.desktop;

import de.pmdungeon.game.Controller.MainController;

public class Main extends MainController {
    public static void main(String[] args) {
        // TODO Auto-generated method stub
        DesktopLauncher.run(new Main());
    }
}

```

### CODE REVIEW

**Comment by cagix** [comment]
is this the "clean" version of #55?

File: build.gradle [hunk 1]
```diff
@@ -1,60 +1,48 @@
...
+    api "com.badlogicgames.gdx:gdx-box2d:$gdxVersion"
+    api "com.badlogicgames.gdx:gdx-freetype:$gdxVersion"
+}
 
-    dependencies {
-        api project(":core")
-        api "com.badlogicgames.gdx:gdx-backend-lwjgl3:$gdxVersion"
-        api "com.badlogicgames.gdx:gdx-backend-lwjgl:$gdxVersion"
-        api "com.badlogicgames.gdx:gdx-platform:$gdxVersion:natives-desktop"
-        api "com.badlogicgames.gdx:gdx-box2d-platform:$gdxVersion:natives-desktop"
```
Comment (cagix) [inline build.gradle:43]: in the old config we used `gdx-box2d-platform:$gdxVersion:natives-desktop`, whereas in the new config there is just `gdx-box2d:$gdxVersion`. is this by design?

File: build.gradle [hunk 2]
```diff
@@ -1,60 +1,48 @@
...
-        api "com.badlogicgames.gdx:gdx-freetype-platform:$gdxVersion:natives-desktop"
```
Comment (cagix) [inline build.gradle:44]: same as with `box2d-platform`

File: build.gradle [hunk 3]
```diff
@@ -1,60 +1,48 @@
...
+    mavenCentral()
+    google()
+    gradlePluginPortal()
 }
 
-project(":desktop") {
-    apply plugin: "java-library"
-    apply plugin: "eclipse"
+dependencies {
+    api group: 'com.google.code.gson', name: 'gson', version: '2.7'
```
[the remark this answers is not in the document]
Comment (cagix) [reply build.gradle:27]: also, in the old config we used a newer version: 2.8.6

File: build.gradle [hunk 4]
```diff
@@ -1,60 +1,48 @@
...
     }
 }
 
-project(":core") {
-    apply plugin: "java-library"
-    apply plugin: "eclipse"
-
-    dependencies {
-        api 'com.google.code.gson:gson:2.8.6'
-        api "com.badlogicgames.gdx:gdx:$gdxVersion"
```
Comment (cagix) [inline build.gradle:55]: this entry is missing in the new config. perhaps we don't need it, but it was generated by the libgdx-builder ...

File: build.gradle [hunk 5]
```diff
@@ -1,60 +1,48 @@
...
-        api "com.badlogicgames.gdx:gdx-box2d:$gdxVersion"
-        api "com.badlogicgames.gdx:gdx-freetype:$gdxVersion"
-        api "com.badlogicgames.gdx:gdx-ai:$aiVersion"
+sourceSets {
```
Comment (cagix) [inline build.gradle:42]: still a bit puzzled why we'd need to declare this ...

File: desktop/build.gradle [hunk 6]
```diff
@@ -1,36 +0,0 @@
-sourceCompatibility = 1.11
-sourceSets.main.java.srcDirs = [ "src/" ]
-sourceSets.main.resources.srcDirs = ["../core/assets"]
-
-project.ext.mainClassName = "de.fhbielefeld.pmdungeon.desktop.DesktopLauncher"
```
Comment (cagix) [inline desktop/build.gradle:5]: shouldn't we declare something like this in the new config, too?

File: desktop/build.gradle [hunk 7]
```diff
@@ -1,36 +0,0 @@
...
-
-task run(dependsOn: classes, type: JavaExec) {
-    main = project.mainClassName
-    classpath = sourceSets.main.runtimeClasspath
-    standardInput = System.in
-    //workingDir = project.assetsDir
-    ignoreExitValue = true
-}
-
-task debug(dependsOn: classes, type: JavaExec) {
```
Comment (cagix) [inline desktop/build.gradle:16]: shouldn't we declare something like this in the new config, too?

File: desktop/build.gradle [hunk 8]
```diff
@@ -1,36 +0,0 @@
...
-    main = project.mainClassName
-    classpath = sourceSets.main.runtimeClasspath
-    standardInput = System.in
-    //workingDir = project.assetsDir
-    ignoreExitValue = true
-    debug = true
-}
-
-task dist(dependsOn: classes, type: Jar) {
```
Comment (cagix) [inline desktop/build.gradle:25]: shouldn't we declare something like this in the new config, too?

**Review by cagix — CHANGES_REQUESTED** [review]
had a quick look. overall: looks promising :)  however, there are quite a few wee questions ...

**Comment by Tobias321** [comment]
@cagix Thanks for the review. Please have a re-look on the new commits. Now we can add a main class and simply call `gradlew run`. (with no including in an IDE...) 🙂

File: build.gradle (hunk 3 above)
[the remark this answers is not in the document]
Comment (Tobias321) [reply build.gradle]: this is just another (or a newer) syntax. i used the old one and a version number property now.

[the remark this answers is not in the document]
Comment (cagix) [reply build.gradle]: yeah, we should either the newer or the older one, but should avoid mixing them ... :)

**Comment by cagix** [comment]
> @cagix Thanks for the review. Please have a re-look on the new commits. Now we can add a main class and simply call `gradlew run`. (with no including in an IDE...) 🙂

so i can finally use my vim again? neat 😀 ... review needs some time, please hold the line ;) ...

**Review by AMatutat — APPROVED** [review]
Works for me. 
Question: All this gradle changes are not relevant for the subproject merge, right?

**Comment by Tobias321** [comment]
> All this gradle changes are not relevant for the subproject merge, right?

The gradle changes are relevant in this pr, because we don't have the projects core and desktop longer.

I tried to make as few changes as possible.

But if we merge this pr, the other pr may have to change...

File: build.gradle (hunk 2 above)
Comment (cagix) [reply build.gradle:44]: still missing?

File: build.gradle (hunk 1 above)
Comment (cagix) [reply build.gradle:43]: done

**Review by cagix — CHANGES_REQUESTED** [review]
still one dependency missing which was originally in build.gradle ...

File: src/de/pmdungeon/desktop/Main.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package de.pmdungeon.desktop;
+
+import de.pmdungeon.game.Controller.MainController;
+
+public class Main extends MainController {
```
Comment (cagix) [inline src/de/pmdungeon/desktop/Main.java:5]: still here?

**Review by cagix — CHANGES_REQUESTED** [review]
`Main` still here?

File: src/de/pmdungeon/desktop/Main.java (hunk 9 above)
Comment (Tobias321) [reply src/de/pmdungeon/desktop/Main.java]: sorry, my fault 😉

Comment (Tobias321) [reply src/de/pmdungeon/desktop/Main.java]: @cagix i could rebase/squash/pick the branch, so that the Main.java commit is not to see in the history ...

**Comment by AMatutat** [comment]
@cagix  @Tobias321 looks good now. Can i merge?

**Comment by Tobias321** [comment]
> @cagix  @Tobias321 looks good now. Can i merge? 

LGTM. I tested on Windows. You can merge.

**Pull request merged**